### DISCUSSION TITLE
Contact listener question

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#78] [Q&A] [answered]
Looking at the description of the Contact Listener methods, I noticed this description for `OnContactRemoved`:

https://github.com/jrouwe/JoltPhysics/blob/314e5baa0a32ba989ea8e20e0a48fadd919ade1d/Jolt/Physics/Collision/ContactListener.h#L106

> Note that we're using BodyID's since the bodies may have been removed at the time of callback.

I am curious if I need to guard against this when using Wasm variant? Can a body be present at `OnContactAdded`, but not in `OnContactRemoved`?

**Comment by jrouwe** [comment, marked as answer]
If you get an OnContactAdded (e.g. box rests on floor) and then after ```PhysicsSystem::Update``` remove the box, then the next time ```PhysicsSystem::Update``` runs you will receive an OnContactRemoved for the box that no longer exists. It is only possible to receive an OnContactAdded and OnContactRemoved for the same contact during a single call to ```PhysicsSystem::Update``` if you run more than 1 collision step (= parameter to the update function).

**Reply by LeXXik** [reply]
I am trying to replicate the use case you described here:

https://jsfiddle.net/LeXXik/sar36vqf/43/

The contact ended closure is called correctly, once a body is removed. However, I am not sure I can detect it correctly on JS side? I think in C++ the `TryGetBody` would return null? In JS, however, it looks like I get an interface that looks valid (albeit it is not, like the user data is no longer correct). Is there a way to query if it is a valid body?

**Reply by LeXXik** [reply]
I am firing a contact removed event in user space, which is driven by that `OnContactRemoved` method. I do have an ability to detect that the entity holding the body is being destroyed, so I can force that event on the frontend framework without Jolt backend having to notify about it. But I still need it for normal cases, when body pairs just naturally stop their contacts. I suppose one way would be to store the body IDs that were scheduled for destruction during this update and compare the IDs from the closure against that list. But, maybe there is a cheaper way.

**Reply by jrouwe** [reply]
> I think in C++ the TryGetBody would return null? In JS, however, it looks like I get an interface that looks valid (albeit it is not, like the user data is no longer correct). Is there a way to query if it is a valid body?

This seems to be another quirk of emscripten. Even though the C++ function returns ```nullptr``` the webidl binder still decides to wrap a Body object around it. Fortunately, you can test if it is null by checking: ```Jolt.getPointer(<body returned by TryGetBody>) == 0```.

**Comment by LeXXik** [comment]
While speaking of Contact Listener, and given your example with a box sitting on floor, will `OnContactPersisted` be called twice? Once for floor and once for box, with same attributes in both calls.

**Reply by jrouwe** [reply]
You will receive only 1 call per contact pair. The order of the objects is specified in [the documentation](https://jrouwe.github.io/JoltPhysics/class_contact_listener.html#aa11e4d0107c2764e1c7e04eef72cb404).

**Reply by LeXXik** [reply]
Ah, I see, it is called once per substep, so can be called multiple times per update. I was wondering why I was getting duplicates. Thanks!
